### ISSUE TITLE
FlexAttention customizability for softmax

### ISSUE THREAD

**Issue by veritas9872** [pytorch-labs/attention-gym#117]
Hello. I am trying to implement the following two features, and I am curious if it is possible to change the inner structure of softmax with FlexAttention.

1. I would like to use a different value for softmax numerical stabilization. Currently, softmax uses exp(value - max_value) for numerical stability. However, this is inefficient because we have to do two memory accesses: once to find the max value and once to get the sum. If the max_value can be set to a constant, this would make the implementation simpler. Also, combined with soft clipping with tanh, there would also be no overflow.
2. Would it be possible to implement Softmax1 from “Attention is off by 1”? I am aware it does not have much effect in practice, but I dislike using an incorrect equation. Also, it might be possible to use a learned bias for the softmax, as implemented indirectly in "Massive Activations in Large Language Models", which could help alleviate the problem of massive activations caused by the need for biases in attention softmax.

I would really appreciate any help you can provide!

**Comment by drisspg** [comment]
I recently added a nice trick that allows you do this: https://github.com/meta-pytorch/attention-gym/blob/main/attn_gym/mods/activation.py
